src/ucs/type/spinlock.h

```diff
@@ -78,9 +77,24 @@ ucs_recursive_spin_is_owner(const ucs_recursive_spinlock_t *lock,
     return lock->owner == self;
 }
 
+static inline int ucs_spin_try_lock(ucs_spinlock_t *lock)
+{
+    return ucs_atomic_cswap32(&lock->lock, UCS_SPINLOCK_FREE,
+                              UCS_SPINLOCK_BUSY) == UCS_SPINLOCK_FREE;
+}
+
 static inline void ucs_spin_lock(ucs_spinlock_t *lock)
 {
-    pthread_spin_lock(&lock->lock);
+    while (!ucs_spin_try_lock(lock)) {
+        while (lock->lock != UCS_SPINLOCK_FREE) {
+            /* spin */
```

Comment:
WDYT adding something like __mm_pause() instruction here (depending on arch)?

Reply:
_mm_pause is not standard extension and may fail to compile on user's system (this header is distributed). we don't add cpu.h to distribution where most of barrier routines are implemented.

OMPI uses similar implementation of spin lock

Reply:
according to some docs, it can improve performance, maybe something like
```c
#ifdef __x86_64__
  asm volatile("pause" ::: "memory")
#endif
```

Reply:
added pause for x86 and read fence for aarch64